_posts/2020-09-20-processing-files-using-java-8-streams.md

```diff
@@ -0,0 +1,382 @@
+﻿---
+title: "Processing Files with Java 8 Streams"
+categories: [java]
+date: 2020-09-20 06:00:00 +1000
+modified: 2020-09-20 06:00:00 +1000
+author: default
+excerpt: ""
+image:
+  auto: 0061-cloud
+---
+
+Streams, introduced in Java 8, use functional-style operations to process data declaratively. The elements of streams are consumed from data sources such as collections, arrays, or I/O resources like files.   
+
+In this article, we'll explore the various possibilities of using streams to make life easier when it comes to the handling of files. We assume that you have a basic knowledge of Java 8 streams. If you are new to streams, you may want to check out [this guide](https://stackify.com/streams-guide-java-8/). 
+
+## Introduction
+
+In the Stream API, there are operations to filter, map, and reduce data in any order without you having to write extra code. Here is a classic example.
+```java
+List<String> cities = Arrays.asList(
+  "London",
+  "Sydney",
+  "Colombo",
+  "Cairo",
+  "Beijing");
+
+cities.stream()
+  .filter(a -> a.startsWith("C"))
+  .map(String::toUpperCase)
+  .sorted()
+  .forEach(System.out::println);
+```
+Here we filter a list of countries, convert to uppercase and sort it before printing the result to the console. 
+
+The output is as below.
+
+```
+CAIRO
+COLOMBO
+```
+
+As the returned streams are lazily loaded, the elements are not read until they are used (which happens when the terminal operation is called on the stream) and thus enhances performance. 
+
+Wouldn’t it be great to apply these SQL like processing capabilities to files as well? How do we get streams from files? Can we walk through directories and locate matching files using streams? Let us get the answers to these questions.
+
+### Code Example
+This article is accompanied by a working code example on [GitHub](https://github.com/thombergs/code-examples/tree/master/core-java/streams/fileswithstreams).
+
+## Getting Started 
+
+Converting files to streams helps us perform many useful operations like counting words in the lines, filtering files based on conditions, removing duplicates from the data retrieved, and so on quite easily. First, let us see how we can obtain streams from files. 
+
+### Building Streams from Files
+
+We can get a stream from the contents of a file line by line by calling the `lines()` method of `Files` class. 
+
+Consider reading from a file bookIndex.txt with the following contents.
+
+**Pride and Prejudice- pride-and-prejudice.pdf  
+Anne of Avonlea - anne-of-avonlea.pdf  
+Anne of Green Gables  - anne-of-green-gables.pdf  
+Matilda  - Matilda.pdf  
+Why Icebergs Float - Why-Icebergs-Float.pdf**
+
+As shown in the example below, the `lines()` method takes the `Path` representing the file as an argument. This method does not read all lines into a `List`, but instead populates lazily as the stream is consumed and this allows efficient use of memory.
+
+#### Example 1
+```java
```

Comment:
Always leave a blank line after a header, otherwise the code example won't be displayed correctly. (I fixed that already)

Reply:
ok